support-frontend/app/services/MParticleClient.scala

```diff
@@ -0,0 +1,169 @@
+    accessToken <- getCachedAccessToken()
+    response <- callProfileAPI(identityId, accessToken.token)
+  } yield parseUserProfile(response)
+
+  private def getCachedAccessToken(): Future[MParticleAccessToken] = {
+    val now = DateTime.now()
+
+    tokenCache.get() match {
+      case Some(cachedToken) if cachedToken.expiresAt.isAfter(now) =>
+        Future.successful(MParticleAccessToken(cachedToken.token))
+      case _ =>
+        getAccessToken().map { case (token, expiresInSeconds) =>
+          val actualExpiryTime = now.plusSeconds(expiresInSeconds)
+          val safeExpiryTime = actualExpiryTime.minus(safetyMargin.toMillis)
+          tokenCache.set(Some(CachedToken(token, safeExpiryTime)))
+          logger.info(
+            s"Cached new mParticle access token, expires at: $safeExpiryTime (${safetyMargin.toSeconds}s safety margin)",
+          )
+          MParticleAccessToken(token)
+        }
+    }
+  }
+
+  private def getAccessToken(): Future[(String, Int)] = {
+    val request = OAuthTokenRequest(
+      client_id = mparticleConfig.clientId,
+      client_secret = mparticleConfig.clientSecret,
+      audience = mparticleConfig.apiUrl,
+      grant_type = "client_credentials",
+    )
+
+    // OAuth token endpoint is on a different domain, so we need to use a custom HTTP client call
+    import com.gu.okhttp.RichOkHttpClient
+    import okhttp3._
+
+    val body = RequestBody.create(
+      MediaType.parse("application/json; charset=utf-8"),
+      request.asJson.noSpaces,
+    )
+
+    val oauthRequest = new Request.Builder()
+      .url(mparticleConfig.tokenUrl)
+      .post(body)
+      .build()
+
+    httpClient(oauthRequest).flatMap { response =>
+      if (response.code() == 200) {
+        io.circe.parser.decode[OAuthTokenResponse](response.body().string()) match {
+          case Right(tokenResponse) =>
+            Future.successful((tokenResponse.access_token, tokenResponse.expires_in))
+          case Left(error) =>
+            val errorMsg = s"Error parsing access token response: ${error.getMessage}"
+            Future.failed(new RuntimeException(errorMsg))
+        }
+      } else {
+        val errorMsg = s"mParticle OAuth returned error: status=${response.code()} body=${response.body().string()}"
+        Future.failed(new RuntimeException(errorMsg))
+      }
+    }
+  }
+
+  private def callProfileAPI(identityId: String, accessToken: String): Future[ProfileResponse] = {
+    val fields = "audience_memberships"
+    val endpoint =
+      s"userprofile/v1/resolve/${mparticleConfig.orgId}/${mparticleConfig.accountId}/${mparticleConfig.workspaceId}"
+
+    val request = ProfileRequest(
+      environment_type = mparticleConfig.apiEnv,
+      identity = Identity(`type` = "customer_id", value = identityId),
+    )
+
+    postJson[ProfileResponse](
+      endpoint = endpoint,
+      data = request.asJson,
+      headers = Map("Authorization" -> s"Bearer $accessToken"),
```

Comment:
unwrap at the last moment
```suggestion
      headers = Map("Authorization" -> s"Bearer ${accessToken.token}"),
```